### PR TITLE
Atomic signals

### PR DESCRIPTION
**Pull request by lucasart**
Rely on well defined behaviour for message passing, instead of volatile. Two
versions have been tested, to make sure this wouldn't cause a slowdown on any
platform.

**v1: Sequentially consistent atomics**

No mesurable regression, despite the extra memory barriers on x86. Even with 15
threads and extreme time pressure, both acting as a magnifying glass:
```
threads=15, tc=2+0.02
ELO: 2.59 +-3.4 (95%) LOS: 93.3%
Total: 18132 W: 4113 L: 3978 D: 10041

threads=7, tc=2+0.02
ELO: -1.64 +-3.6 (95%) LOS: 18.8%
Total: 16914 W: 4053 L: 4133 D: 8728
```
**v2: Acquire/Release semantics**

This version generates no extra barriers for x86 (on the hot path). As expected,
no regression either, under the same conditions:
```
threads=15, tc=2+0.02
ELO: 2.85 +-3.3 (95%) LOS: 95.4%
Total: 19661 W: 4640 L: 4479 D: 10542

threads=7, tc=2+0.02
ELO: 0.23 +-3.5 (95%) LOS: 55.1%
Total: 18108 W: 4326 L: 4314 D: 9468
```
As suggested by Joona, another test at LTC:
```
threads=15, tc=20+0.05
ELO: 0.64 +-2.6 (95%) LOS: 68.3%
Total: 20000 W: 3053 L: 3016 D: 13931
```
Note that, on ARM and PPC, acquire/release will generate barriers, unlike x86.
And we don't have ARM and PPC machines on fishtest. This is one of the reson for
testing version 1 as well. The theoretical cost of these barriers is not
mesurable, as shown by v1 tests.

Obviously, we go with v2.

No functional change.

### CODE REVIEW

**Comment by lucasart** [comment]
@zamar: same one as before, rewritten and reopened (Marco's rebase broke the previous branch).

**Comment by glinscott** [comment]
Nice, thanks!  That's a really nice cleanup.  Approved.

**Comment by mcostalba** [comment]
@lucasart can you please better explain what's the point of this patch, I mean, to explain in practical terms. Thanks.

**Comment by mcostalba** [comment]
std::memory_order_acquire and friends are used to replace locks, not volatile.

They are used in a typical producer/consumer pattern to sync the threads in this way:

Consumer:
   p = guard.load(std::memory_order_acquire)
   if (p) {  resource.consume();  guard.store(0, std::memory_order_release); }

Producer:
   p = guard.load(std::memory_order_acquire)
   if (!p) {  resource.produce();  guard.store(1, std::memory_order_release); }


But they are not intended to replace volatile in general case, and in our specific case volatile is _the_ correct term to use because we don't have any synchronization to do,  just detect if a variable has been changed by an external thread. Please note the 2 threads are _not_ synchronized nor we want them to be synchronized because stop signal can and rises at any random time: the fact that the second thread reads stop value some instructions after it has been rises it is totally harmless and indeed is intrinsic in the randomness of the rising event.

We use volatile here for just a single reason: to force the compiler to read the variable and this is _exactly_ the semantic for which volatile was designed. 

So this is one of the (few) real correct usage scenarios of volatile keyword!

**Comment by mcostalba** [comment]
@lucasart eventually what you really want is for teh signakl to be a sig_atomic_t

http://en.cppreference.com/w/c/program/sig_atomic_t

Indeed volatile allow uses of sig_atomic_t variables in signal handlers, see https://en.wikipedia.org/wiki/Volatile_(computer_programming)

**Comment by zamar** [comment]
We don't have any signal handlers in SF. And I don't expect us to have any. I'm confused.

**Comment by lucasart** [comment]
@mcostalba: An atomic write with release is visible immediately to anyone reading it with acquire. Volatile offers zero guarantee. With volatile, the CPU could do a lazy write to a cache, and it may take a long time before it gets flushed into global memory, or to some cache level shared by everyone. There is no time limit before the volatile write becomes globally visible.

In *practice* atomics with acquire/release will generate the same code on x86 as relaxed, which will probably generate the same code as volatile. But the point is that atomics give you a.guarantee from the standard that no compiler or weird platform will ever be able to violate that. Writing a signal mt be immediatly and globally visible.

**Comment by lucasart** [comment]
@mcostalba: your link on sig_atomic is from C99, not C++11. Once you click on C++ documentation, it says:
```
 Until C++11, which introduced std::atomic and std::atomic_signal_fence, about the only thing a strictly conforming program could do in a signal handler was to assign a value to a volatile static std::sig_atomic_t variable and promptly return
```
See? Your sig_atomic is old and busted. The C++11 way is to use `std::atomic`.

**Comment by mcostalba** [comment]
@lucasart ok, but in this case why don't you use atomic_flag that is
guaranteed to be lockfree (we really don't need syncroniztion)

http://en.cppreference.com/w/cpp/atomic/atomic_flag

**Comment by mcostalba** [comment]
Ok, sorry I see atomic_flag cannot be read (does not support load
operation), so is not useful for us.

**Comment by mcostalba** [comment]
Ok, I have read more deeply the atomic documentation and specifically
std::memory_order

http://en.cppreference.com/w/cpp/atomic/memory_order

Now, if you want to implement what you advertise: " An atomic write with
release is visible immediately to anyone reading it with acquire" then the
only way is to use memory_order_seq_cst with possible slowdown.

In particularly memory_order_acquire and memory_order_release enforce
ordering only for the instruction **in the same thread** but there is no
guarantee of **when** the other threads will see the change.

**Comment by zamar** [comment]
@mcostalba: acquire/release pair achieves that:

memory_order_acquire: ... This ensures that all writes in other threads that release the same atomic variable are visible in the current thread.

memory_order_release: This ensures that all writes in the current thread are visible in other threads that acquire or the same atomic variable and writes that carry a dependency into the atomic variable become visible in other threads that consume the same atomic.

**Comment by mcostalba** [comment]
@joona we want to ensure that are visible _immediately_,  for what I
understand memory_order_acquire and memory_order_release simply not reorder
the writes /reads to cache / RAM so that for instance on Intel platform
that has strong ordering, these operation are simply the usual load/store
asm instructions.

But we don't have _any_ guarantee to _when_ the write buffer is flushed to
RAM.

So that a core can write stop = 1 and this writes remains pending for a not
defined time before to be flushed, the guarantee is that also all teh
following writes/loads are pending, but there is _no_ guarantee when the
data will be visible to all other threads.

**Comment by lucasart** [comment]
@mcostalba: You're right. It should be atomic_flag. In fact, our spinlock should also use atomic_flag. Can you do the necessary changes ? I am on holiday, and wont be able to code for a week.

**Comment by zamar** [comment]
@mcostalba: OK. There is no guarantee about the timing... Maybe the atomic_flag is indeed the way to go...

**Comment by mcostalba** [comment]
@zamar the way to go IMO is to use:

Signals.stop.load(std::memory_order_relaxed)


Signals.stop.sore(true, std::memory_order_seq_cst)


When we read we don't enforce any ordering because event is
intrinsically asynchronous, instead when we write we want that the new
value is immediately updated to all threads.


This is the most correct and also the fastest because the read
operations are on the hot path, instead the write operation happens
only once per move.


I have checked Lucas's patch but I see he has updated only very few
places, there are still many occurrence of 'stop' and otehr signals
that are not updated to atomic load/store version.

**Comment by mcostalba** [comment]
BTW I am not sure we can get rid of volatile, see store() signatures

http://en.cppreference.com/w/cpp/atomic/atomic/store

There is also the 'volatile' overload...this is a bit scary...it means that
without volatile we can have

if (Signals.stop.load(std::memory_order_relaxed)) {.... do something .... }

else if (Signals.stop.load(std::memory_order_relaxed)) { .....do something
else }  // This could be optimized away by the compiler!

**Comment by lucasart** [comment]
@mcostalba: the default is seq_cst which is stronger than both acquire and release. that is what you should always use, unless you have a reason not to. only on the hot path is where you want acquire/release to save memory barriers. this is exactly what i am doing.

There is a guarantee about timing. In the words of the standard "a reasonable amount of time". Volatile OTOH is allowed to never flush the buffer and the core would just read/write from its own cache w/o anything being visible to other threads for a long time.

Nothing will give you any stronger guarantee than atomics.

**Comment by zamar** [comment]
@mcostalba:

Signals.stop = true

defaults to "Signals.stop.store(true, std::memory_order_seq_cst)"

**Comment by mcostalba** [comment]
@zamar @lucasart

Ok, we are almost there :-)

1) I suggest

Signals.stop.load(std::memory_order_relaxed)


for reading, mainly because we don't need consumer-producer pattern, we
don't have to synchronize anything and this flag seems to me the correct
one in our scenario (as a side effect is also the fastest). Of course for
setting we want to use std::memory_order_seq_cst.

2)  I think volatile is needed anyhow to avoid compiler to optimize out a
second read of the flag in the same function (as is the case with 'stop'
in search).  I see that atomic have a 'volatile' overload this seems to
imply that volatile is still needed to aggressive prevent compiler
optimizations.

**Comment by zamar** [comment]
@mcostalba: I'd be very surprised if compilers were allowed to optimize out atomic reads... That would kind of defeat the whole purpose of 'atomics'.

**Comment by mcostalba** [comment]
Nevertheless standard library foresee a volatile overload for atomic
functions. What is the reason for this?

**Comment by zamar** [comment]
Discussed here: http://stackoverflow.com/questions/2479067/why-is-the-volatile-qualifier-used-through-out-stdatomic

**Comment by zamar** [comment]
Okay, this gets very tricky, but my understanding is that volatile overloads are provided, because some variables may be both atomic and volatile.

volatile = the value may change at any time (because something outside the program changes it)
atomic = guarantees atomic reads/writes to the variable.

Unfortunately the standard isn't too wordy about the very nature of atomicity and allowed/disallowed computer optimizations. There is the following note though about memory_order_relaxed (29.3):

[quote]
Note:
Atomic operations specifying memory_order_relaxed are relaxed with respect to memory ordering.
Implementations must still guarantee that any given atomic access to a particular atomic object be indivisible with respect to all other atomic accesses to that object.
— end note
[/quote]

**Comment by zamar** [comment]
I can't find anywhere the statement that you're looking for (that compilers can't optimize way atomic loads).

However if this was the case, then it would render atomics mostly useless, e.g. the following spin lock code that we have wouldn't work:

  std::atomic_int lock;

  void acquire() {
      while (lock.fetch_sub(1, std::memory_order_acquire) != 1)
          while (lock.load(std::memory_order_relaxed) <= 0)
              std::this_thread::yield(); // Be nice to hyperthreading
  }

Because compiler could optimize it into:

  void acquire() {
      while (lock.fetch_sub(1, std::memory_order_acquire) != 1)
      {
          int x = lock.load(std::memory_order_relaxed)
          while (x <= 0)
              std::this_thread::yield(); // Be nice to hyperthreading
  }

**Comment by mcostalba** [comment]
@joona yes, I understand, but in case some compiler does, we really
introduce  very subtle and almost impossible to find bug.

This practically what I have understood:

volatile = force the compiler to access the variable

atomic = variable is read/write in an atomic way relative to other threads,
i.e. no other thread access the variable during read/write

std::memory_order_seq_cst = makes a write to the variable to be immediately
visible to all the threads.


IMO we need all 3, at least we need all the 3 properties that they enforce.

**Comment by zamar** [comment]
Unfortunately I don't think that the following holds:
std::memory_order_seq_cst = makes a write to the variable to be immediately visible to all the threads.

It gives you a guarantee that changes to atomic variables will appear in the same order for all threads, but again there is no guarantee about how quickly this will happen (other than reasonable time).

**Comment by snicolet** [comment]
It seems that Herb Sutter deals with exactly this issue (a flag set by a master thread to stop slave threads) at 1:17:10 in the second video (Atomic Weapons 2/2) which Joona referred to some days ago:
https://isocpp.org/blog/2013/02/atomic-weapons-the-c-memory-model-and-modern-hardware-herb-sutter

His conclusion is that stop can just be an atomic, with stop.load() using a relaxed memory order.

**Comment by snicolet** [comment]
About (optimization and) non-optimization of atomics: the documentation for the LLVM compiler is online at http://llvm.org/docs/Atomics.html#atomic-instructions . It describes precisely what optimizations the front-end writers for this compiler are allowed to implement.

**Comment by mcostalba** [comment]
It does not compile for me on mingw/Windows, I will push a fix.

**Comment by mcostalba** [comment]
I have pushed a commit above Lucas's one.

https://github.com/mcostalba/Stockfish/commit/atomic_signals

Lucas, please evaluate it and eventually update your pull request. Thanks.

**Comment by lucasart** [comment]
so what is the gain in doing this unorthodox load relaxed / store seq_cst, compared to the normal load acquire / store release ?

you will note that we only test on x86, where acquire/release/relaxed are the same thing, so we are taking a risk in trying to be clever with relaxed. a risk of subtle breakage on ARM where we dont test, for no gain…

**Comment by mcostalba** [comment]
The gain is that this is _the_ correct way to do this.

 load acquire / store release does not apply in this case.

The Herb Sutter video posted by Stephane explain this well.

**Comment by zamar** [comment]
I think both ways should work :-)

seq_cst/relaxed is probably theoretically slightly better, as there is no need to synchronize threads, but in practice it probably doesn't matter.

**Comment by mcostalba** [comment]
@lucasart the only risk of subtle bugs we are taking is the removal of volatile qualifier that we "assume" should work and compiler should not optimize away atomic access. But this is not documented in any place.

**Comment by lucasart** [comment]
i have to close this, and resolve merge conflicts again, as lazy was merged into master...
